feat(inccommand): allow redrawing during cmdpreview

theofabilous opened the pull request:
**Problem**: The current `inccommand` implementation works by forcing a redraw right after applying changes and immediately restoring everything without a subsequent redraw, thereby relying on ephemeral screen state (the "preview" visible to the user does not represent the actual state of buffers/windows). This requires disabling redraws while command preview is active (#9777, #9783), which suppresses unrelated UI updates (#20463, #27950), complicates/prevents redrawing floating windows (#28510), and causes rendering issues for GUIs (#24802). In cases where a redraw is forced (e.g. editor resize), the command preview is cleared. Similarly, `<C-g>` and `<C-t>` during `:substitute` clear the preview.

**Solution**: Remove reliance on temporary screen state in command preview, actually keep the buffer changes after applying them. Only restore state when needed (e.g. before re-updating the preview, when leaving cmdline, ...). Allow redrawing during preview, and refresh command preview when the editor is resized or a scroll occurs.

---

> Helpful background: #28510. Notably, it highlights why refreshing command preview is required when scrolls/resizes occur (even when the preview state is persistent). In brief, commands may choose to apply previews only within the visible viewport, so when more lines are revealed the preview may need to be executed again. (If this weren't the case, the editor would basically brick on each keypress after `:%substitue/` for a sufficiently large file).

<details><summary><b>NOTE</b> about commit history</summary>

The commit history is kind of wonky, I tried to organize changes in a way that would make the implementation changes easier to follow (incrementally/chronologically) at the expense of a clean/logical commit sequence in its own right. When this PR gets closer to being ready and others have taken a look at it, I'd like to squash and rearrange some of these changes.

</details>

---

# Implementation

> **NOTE**: In this section, I often use the term "preview callback" or something similar, by which I mean the "per-command" previewing function. So for user-commands thats the `preview = ...` field in the `opts` arg of `nvim_create_user_command` (`:help :command-preview`), and for builtin functions that's `CommandDefinition.cmd_preview_func` (currently only exists for `:substitute`).

The changes in this PR can be roughly organized in 4 steps/categories, which loosely follow the commits chronologically:

1. Refactor the existing cmdpreview logic to ease the new implementation, without yet changing observable behaviour (i.e. still use the `apply + redraw + teardown/restore + disable redrawing` approach). The original approach only ever executes the command preview routines entirely and atomically, all in one go. The new approach requires executing different parts of this routine independently. Additionally, the preview state now needs to be kept alive across calls to update functions and may also be called externally (like, not within the "regular" command line event handling functions). This requires hoisting the state object to a global.

<details>

- split `cmdpreview_prepare()` in two: `cmdpreview_save()` and `cmdpreview_prepare()`. `save()` handles saving buffer information (undo, changedtick, ...), while `prepare()` sets things up only needed for the duration of the next preview execution (ensures undolevels are such that we can subsequently rollback changes, temporarily disable cursorline and incsearch, ...). Reason: later, we'll often want to only call `prepare()` without saving again.
- split `cmdpreview_restore_state()` in two, moving out memory deallocations to `cmdpreview_destroy()`. We'll eventually want to keep saved information around, but we still need to restore buffers before re-executing preview changes, so need to split restoration and deallocation.
- Use a global `CpInfo` object instead of one local to `cmdpreview_may_show()`. Move some of the local variables in that function to fields in `CpInfo` if we'll eventually need to keep track of them.
- extract part of the initial logic from `cmdpreview_may_show` to `cmdpreview_may_init`/`cmdpreview_check_cmdline`, which handle checking if the command is previewable and, if applicable, preparation for showing preview.
- reorganize where/when autocommands are blocked and messages are disabled. Before this wasn't a concern because all the subroutines where only ever called by `cmdpreview_may_show()`, so that function could simply block autocmds at the beginning and unblock them at the end. Eventually, we'll be calling some of those functions individually but still want to block events and error reporting.

</details>

At this point, the high level logic is the same: `cmdpreview_may_show()` handles the full init->show->destroy cycle at once.

2. Don't immediately tear down preview state. Now, `cmdpreview_may_show()` no longer unconditionally completes the full preview cycle. On preview success, the buffer changes are no longer reverted immediately. The `CpInfo` object is not destroyed and thus still holds the saved buffer information. On the next call, the buffer information doesn't need to be saved: instead, the buffers are just restored using the existing saved information before re-applying the preview changes.

<details>

- add `cmdpreview_transition_begin`/`_end` functions. These handle maintaining invariants that I found hard to keep track of manually. When entering the preview handling functions, we may or may not be already previewing, and when exiting the cmdpreview handling functions, we may or may not need to restore buffers, free memory, and/or close windows (for example, depending on whether the preview callback succeeded). The transition functions handle a lot these things.
- we previously didn't always call `cmdpreview_may_show` when the command line changes, this no longer works because the preview may still be active; skipping the cmdpreview updates would then keep those changes active. So, if the cmdline was updated but cmdpreview shouldn't be shown (e.g. when not interactive or input chars are available), call `cmdpreview_did_not_show` which handles cleanup, if applicable.

</details>

3. Handle buffer and window changes/invalidations between preview updates. Now that the changes are kept around, re-executing preview routines when preview is already active isn't as simple: buffers and windows may have been closed, so some of the saved buffer/window information may no longer be valid, and there may be new windows/buffers that now need to be saved.

<details>

- Use `bufref_T` instead of bufs for saved buffer information, so that we can check if the information refers to a invalid buffer. Free invalid buffer information without restoring them in these cases
- Store saved information in hashmaps instead of vecs: when preparing for updates, we want to go through open windows/buffers and check if we've already saved them. See commit message of 7d54e8b
    - **NOTE**: this is possibly less efficient in general unless you have a lot of buffers/windows open, but I'm going with naive/"obvious" solutions for the time being

</details>

4. Finally, allow redrawing while command preview is active. Also update cmdpreview when a resize or scroll occurs, if needed (see note above and/or #28510 for why the latter is required)

---

# Considerations/open questions

Now that preview changes are persistent, a lot more bookkeeping is required. Specifically, buffer and window information can arbitrarily change in between preview updates, and I'm not sure how to handle all such changes.
I have a few specific concerns/questions that I'll add in the comments when I get around to properly wording them.

Here's a rough, quick, likely non-exhaustive list of these concerns:

- What exactly is a preview function allowed to do? I think this should be documented, both for users' and devs' sake.

- In the spirit of playing nice with external UIs, I think there needs to be a way to mark windows as ignored by command preview (if there isn't already), or maybe just have a class of window/buffer option combinations that should be ignored by command preview (e.g. `nomodifiable`+`bufhidden=hide`+`buftype=nofile`+is a floating window+...)?
    - For now, nothing is ignored, but to allow UI widget modifications while cmdpreview is active, the following best effort "heuristic" approach is used (see https://github.com/neovim/neovim/pull/28856/commits/d5ea89cf300fdcd13bc0868cc7202ca168731241):
        - after the preview callback, mark all unchanged buffers (reason: we may "assume" that buffers not modified by the preview callback are more likely to be UI widget-ish)
        - on the next preview update, if a marked buffer has changed, don't attempt to restore it (reason: ui widgets may have been modified in the meantime, restoring them makes no sense and would break e.g. externalized cmdlines during preview)

- What/how do we handle the "in between preview updates" part? Preview changes are actually persistent now, and things can potentially be "messed up" in between consecutive calls to a preview callback (e.g. autocommands called after cmdline changed right before cmdpreview is executed, or a timer callback, ...). Specifically:
    - what if a buffer was changed during this time (possibly *after* being modified by cmdpreview, but *before* those changes were reverted)?

luukvbaal reviewed (commented):
Thanks for working on this! It seems quite strange to me to cram this into the redraw routines. If there is a need to refresh the inccommand window(s) dynamically, perhaps it makes sense to add a dedicated API for that?

EDIT: Oops, meant to add this to the change in `drawscreen.c`. Ah I remember now the changed viewport thing(scroll/resize), is this mainly for that? Could you add a docstring to `cmdpreview_may_refresh()` that explains why it is needed?

theofabilous committed:
refactor: split cmdpreview funcs, use global var

theofabilous committed:
refactor: move previewable check to own function

theofabilous committed:
refactor: move msg silencing and autocmd blocking

theofabilous committed:
feat: don't immediately tear down cmdpreview stuff

theofabilous committed:
fix(test): outdated `inccommand` tests

A couple of tests are dependent on the implementation details of command
preview (notably, the timing/ordering of buffer restoration). Update
these tests to be compatible with the new implementation.

theofabilous committed:
feat(icm): lock `icm=split` preview window/buffer

Add `b_cp_locked` buffer field that indicates if the buffer is locked by
cmdpreview. It acts kind of like `b_locked` + `b_locked_split`, and
`w_locked` + `w_p_wfb` for a window that contains it.

When set to `BCP_FULL_LOCKED`, the buffer cannot be closed nor entered,
and windows that contain it cannot be closed, split, or change
their buffer.

When set to `BCP_HALF_LOCKED`, all restrictions apply except that the
buffer *can* be unloaded.

theofabilous committed:
feat: update saved win/buf info between previews

Now that the preview state is not destroyed immediately after showing
it, the buffer and window info isn't restored immediately either, but kept
in the cmdpreview info structure. In between previews, buffers, windows
and tabs may have changed and/or closed, so the saved information must
be kept in sync, and care must be taken not to attemp to restore state
for a non-existent buffer/window.

The major changes are as follows:

1. Store saved buffer/window info in hash maps instead of lists, because
   we're often checking if a certain buffer/window is already saved.

2. Now a `bufref_T` is kept instead of just the buffer pointer; we need
   to check if the buffer is still valid (otherwise, the buffer may have
   been destroyed and the pointed-to memory may have been reused). A
   similar approach is used with the window info, we store both the
   window pointer and its handle, and check if the window is valid and
   has the same handle as before.

3. Before restoring state, go through all saved buffers and windows and
   make sure they are still valid. If they aren't, remove and free their
   saved information. This accounts for buffers and windows that are
   *removed* in between previews.

4. Now `cmdpreview_save()` is also called when the preview is already
   initialized; instead of clearing the buf_info/win_info fields, it
   adds info into them if not already stored. This is used to account
   for buffers and windows that are *added* in between previews.

theofabilous committed:
feat(icm): allow redrawing during command preview

theofabilous committed:
feat(icm): refresh preview on scroll and resize

theofabilous committed:
fix(icm): restore unchanged bufs after cmdpreview

Problem: Buffers for external ui widgets (such as external cmdlines)
are restored to the initial save state on every preview update, even if
they were only modified in between previews.

Solution: Mark buffers that are unchanged by the preview callback, don't
restore them, and instead update their save state on the next preview
update, so that subsequent restorations (if any) will use this new save
state in place of the initial state collected on the first preview.

justinmk commented:
> Now that preview changes are persistent, a lot more bookkeeping is required.

Yeah, 500 lines of new C code isn't a feasible way to approach this. Think about leveraging UI events or at least shelling out to Lua.

Feel free to rebase and reopen if a lower-cost approach can be found.
